hawtio-system/src/test/java/io/hawt/util/StringsTest.java

```diff
@@ -43,6 +44,56 @@ public void split() {
         }
     }
 
+    @Nested
+    class PropertiesTest {
+
+        @Test
+        public void oneLevelSystemProperties() {
+            System.setProperty("hawtio.what", "world");
+            assertThat(Strings.resolvePlaceholders(null), nullValue());
+            assertThat(Strings.resolvePlaceholders("Hello"), is("Hello"));
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.what}!"), is("Hello world!"));
+            assertThat(Strings.resolvePlaceholders("${hawtio.what}"), is("world"));
+            assertThat(Strings.resolvePlaceholders("${hawtio.what}, hello ${hawtio.what}!"), is("world, hello world!"));
+        }
+
+        @Test
+        public void valuesWithPlaceholdersAreNotResolved() {
+            System.setProperty("hawtio.what", "${world}");
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.what}!"), is("Hello ${world}!"));
+        }
+
+        @Test
+        public void mismatchedPlaceholders() {
+            System.setProperty("nestedProperty", "what");
+            System.setProperty("hawtio.what", "world");
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.what!"), is("Hello ${hawtio.what!"));
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.${nestedProperty}!"), is("Hello ${hawtio.what!"));
+        }
+
+        @Test
+        public void nestedSystemProperties() {
+            System.setProperty("doubleProperty", "hawtio.what");
+            System.setProperty("nestedProperty", "what");
+            System.setProperty("hawtio.what", "world");
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.${nestedProperty}}!"), is("Hello world!"));
+            assertThat(Strings.resolvePlaceholders("${hawtio.${nestedProperty}}"), is("world"));
+            assertThat(Strings.resolvePlaceholders("${hawtio.${nestedProperty}}, hello ${hawtio.${nestedProperty}}!"), is("world, hello world!"));
+            assertThat(Strings.resolvePlaceholders("${doubleProperty}"), is("hawtio.what"));
+            assertThat(Strings.resolvePlaceholders("${${doubleProperty}}"), is("world"));
+            assertThat(Strings.resolvePlaceholders("${${${doubleProperty}}}"), is("${world}"));
+        }
+
+        @Test
+        public void recursiveSystemProperties() {
+            System.setProperty("nestedProperty", "what");
+            System.setProperty("hawtio.what", "hawtio.${nestedProperty}");
+            assertThat(Strings.resolvePlaceholders("Hello ${hawtio.${nestedProperty}}!"), is("Hello hawtio.${nestedProperty}!"));
```

Comment:
why isn't it resolved to `Hello hawtio.what!`?

Reply:
First, nested `${nestedProperty}` is resolved to `what`, but because it's resolved inside other pair of `${...}` the value is another key, so we have `Hello ${hawtio.what}!`, but the value of `hawtio.what` system property (which is literally `hawtio.${nestedProperty}`) is not resolved recursively (we can do it if needed), but simply added to the output.